src/Pos/Delegation/Logic/VAR.hs

```diff
@@ -512,15 +536,10 @@ removeNoLongerRichmen ::
        )
     => EpochIndex
     -> m SomeBatchOp
-removeNoLongerRichmen (EpochIndex 0) = pure mempty
 removeNoLongerRichmen newEpoch = do
-    let getRichmen e =
-            toList <$>
-            lrcActionOnEpochReason e "removeNoLongerRichmen" LrcDB.getRichmenDlg
-    oldRichmen <- getRichmen (newEpoch - 1)
-    newRichmen <- getRichmen newEpoch
-    let edgeActions = map DlgEdgeDel $ oldRichmen \\ newRichmen
-    calculateTransCorrections $ HS.fromList edgeActions
+    noLongerRichmen <- getNoLongerRichmen newEpoch
+    let edgeActions = map DlgEdgeDel noLongerRichmen
+    calculateTransCorrections (HS.fromList edgeActions)
```

Comment:
This is weird. You should return edge actions with trans corrections. Edge actions modify db (1), while trans corrections affect transitive and transitive reverse (2-3).